### PR TITLE
Enforce client protocol checking on auth flows

### PR DESCRIPTION
**Pull request by josh-cain**
Currently authentication flows do not validate that the client's declared protocol is in fact the LoginProtocol that's being used.  This change updates the flows such that they're checked.

I followed suite with the surrounding code, and threw an AuthenticationFlowException, however I quickly discovered that the OIDC /token endpoint did not handle these gracefullly.  Added a little code to clean it up there (and it looked like the Authorization endpoint did some pokemon handling, so it should be caught there too), but can't say that I'm familiar enough with it to mark it as comprehensively covered.

### CODE REVIEW

**Comment by cainj13** [comment]
Jira issue here: https://issues.jboss.org/browse/KEYCLOAK-6718

File: services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java [hunk 1]
```diff
@@ -503,10 +504,17 @@ public Response resourceOwnerPasswordCredentialsGrant() {
...
+        try {
+            Response challenge = processor.authenticateOnly();
+
+            if (challenge != null) {
+                cors.build(httpResponse);
+                return challenge;
+            }
+        } catch (AuthenticationFlowException e) {
+        	// Can occur if the wrong protocol is used, user is under brute force protection, etc.
+            event.error(Errors.IDENTITY_PROVIDER_ERROR);
```
Comment (stianst) [inline services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java]: The error used here is not correct.

File: services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java [hunk 2]
```diff
@@ -503,10 +504,17 @@ public Response resourceOwnerPasswordCredentialsGrant() {
...
+            throw new CorsErrorResponseException(cors, OAuthErrorException.ACCESS_DENIED, "Access denied", Response.Status.UNAUTHORIZED);
```
Comment (stianst) [inline services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java]: This is masking the real problem and just using a generic access denied error which doesn't seem correct either

**Review by stianst — CHANGES_REQUESTED** [review]
There's no test to check wrong protocol isn't allowed. There's also a few tests that are now failing.

File: services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java (hunk 2 above)
Comment (cainj13) [reply services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java]: Sure.  So this wasn't there before, and the token Endpoint today just throws a 500 if an AuthenticationFlowException occurs during the user auth process.  I can just put the behavior back to be what it is today...

File: services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java (hunk 1 above)
Comment (cainj13) [reply services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java]: Same as below.  Can just remove the catch if that's more preferable.

**Comment by cainj13** [comment]
Can certainly add the test and address the failures, I just assumed there were some unstable tests, as my Travis builds fail more frequently than pass these days :(

**Comment by stianst** [comment]
@cainj13 bump

**Comment by cainj13** [comment]
@stianst  was waiting on guidance from question above ^.  So just put things back like they were and remove the catch for AuthenticationFlowException's such that they just cause a 500 from this path?

**Comment by stianst** [comment]
I think that's the best. If we want to address the returned status code that can be covered by a separate issue/PR.

**Comment by stianst** [comment]
I'm going to close this PR due to lack of updates. Also, see my comment in the JIRA. I don't think this issue is limited to authentication flows, but is most likely an issue to all protocol endpoints.

**Pull request closed without merging**